src/main/java/org/terasology/web/webSocket/WebSocketHandler.java

```diff
@@ -72,28 +78,50 @@ public void onWebSocketError(Throwable cause) {
         logger.error("Error", cause);
     }
 
-    private void handleClientMessage(ClientMessage clientMessage) {
+    private void handleClientMessage(ClientToServerMessage clientMessage) {
         switch(clientMessage.getMessageType()) {
             case AUTHENTICATION_REQUEST:
                 trySendResult(jsonSession.initAuthentication()); //send server handshake hello
                 break;
             case AUTHENTICATION_DATA:
                 trySendResult(jsonSession.finishAuthentication(clientMessage.getData())); //process client handshake hello
                 break;
-            case RESOURCE:
-
+            case RESOURCE_REQUEST:
+                handleResourceRequest(clientMessage.getData());
         }
     }
 
-    private void trySend(String message) {
+    private void trySend(ServerToClientMessage message) {
         try {
-            getSession().getRemote().sendString(message);
+            getSession().getRemote().sendString(GSON.toJson(message));
         } catch (IOException e) {
             logger.warn("Unable to send message!", e);
         }
     }
 
+    private void trySendResult(ActionResult result, String resourceName) {
+        trySend(new ServerToClientMessage(ServerToClientMessage.MessageType.ACTION_RESULT, resourceName, result.toJsonTree(GSON)));
+    }
+
     private void trySendResult(ActionResult result) {
-        trySend(result.toJsonString(GSON));
+        trySend(new ServerToClientMessage(ServerToClientMessage.MessageType.ACTION_RESULT, result.toJsonTree(GSON)));
+    }
+
+    private void handleResourceRequest(JsonElement requestMessage) {
+        ResourceRequestClientMessage deserializedMessage;
+        try {
+            deserializedMessage = GSON.fromJson(requestMessage, ResourceRequestClientMessage.class);
+        } catch (JsonSyntaxException ex) {
+            trySendResult(new ActionResult(ex));
+            return;
+        }
+        String resourceName = deserializedMessage.getResourceName();
```

Comment:
I would put this part inside the `try { .. }` or in a separate method that is called from the try block. This makes it easier to read.